### PR TITLE
Gamepad input using GLFW

### PR DESCRIPTION
**Pull request by Ingrater**
* Add GLFW support on windows (optional, default off)
* Implement Gamepad input using GLFW
* Add central access point for getting an controller input device. (ezControllerInput)

### CODE REVIEW

File: Code/Engine/Core/System/ControllerInput.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+#include <Core/CoreDLL.h>
+
+class ezInputDeviceController;
+
+class EZ_CORE_DLL ezControllerInput
+{
+public:
+  // \brief Returns the global controller input device. May be nullptr.
+  static ezInputDeviceController* GetDevice();
```
Comment (jankrassnigg) [inline Code/Engine/Core/System/ControllerInput.h:14]: Could add a "HasDevice" function for non-null checks.

File: Code/Engine/Core/System/Implementation/glfw/ControllerInput_glfw.inl [hunk 2]
```diff
@@ -0,0 +1,227 @@
+#include <Core/Input/DeviceTypes/Controller.h>
```
Comment (jankrassnigg) [inline Code/Engine/Core/System/Implementation/glfw/ControllerInput_glfw.inl:1]: Why is there an empty cpp file ?

File: Code/Engine/Core/System/Implementation/glfw/ControllerInput_glfw.inl [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+void ezControllerInputGlfw::SetDeadZone(const char* szButton)
+{
+  ezStringBuilder s;
+
+  for (ezInt32 i = 0; i < MaxControllers; ++i)
+  {
+    s.Format("controller{0}_{1}", i, szButton);
+    ezInputManager::SetInputSlotDeadZone(s.GetData(), 0.23f);
```
Comment (jankrassnigg) [inline Code/Engine/Core/System/Implementation/glfw/ControllerInput_glfw.inl:115]: what's the purpose of this ?

File: Code/Samples/Asteroids/GameState.cpp [hunk 4]
```diff
@@ -71,12 +69,10 @@ void AsteroidGameState::BeforeWorldUpdate()
 
 void AsteroidGameState::ConfigureInputActions()
 {
-#if EZ_ENABLED(EZ_PLATFORM_WINDOWS)
-  ezInputDeviceXBox360::GetDevice()->EnableVibration(0, true);
-  ezInputDeviceXBox360::GetDevice()->EnableVibration(1, true);
-  ezInputDeviceXBox360::GetDevice()->EnableVibration(2, true);
-  ezInputDeviceXBox360::GetDevice()->EnableVibration(3, true);
-#endif
+  ezControllerInput::GetDevice()->EnableVibration(0, true);
```
Comment (jankrassnigg) [inline Code/Samples/Asteroids/GameState.cpp]: Wouldn't you need to wrap all of these calls in a check? Unless you provide a fallback NULL implementation for controller devices.

File: Code/ThirdParty/glfw/finalize.cmake [hunk 5]
```diff
@@ -11,4 +11,10 @@ if(TARGET glfw)
 
 
     set_property(GLOBAL APPEND PROPERTY "EXPORT_PROJECTS" glfw)
+    
+    ez_set_project_ide_folder(glfw "Code/ThirdParty/GLFW")
+    ez_set_project_ide_folder(update_mappings "Code/ThirdParty/GLFW")
```
Comment (jankrassnigg) [inline Code/ThirdParty/glfw/finalize.cmake:16]: what is this ?

**Review by jankrassnigg — APPROVED** [review]

File: Code/Engine/Core/System/Implementation/glfw/ControllerInput_glfw.inl (hunk 2 above)
Comment (Ingrater) [reply Code/Engine/Core/System/Implementation/glfw/ControllerInput_glfw.inl:1]: Accident, should not be there.

File: Code/Engine/Core/System/Implementation/glfw/ControllerInput_glfw.inl (hunk 3 above)
Comment (Ingrater) [reply Code/Engine/Core/System/Implementation/glfw/ControllerInput_glfw.inl:115]: I copied the SetDeadZone code straight from the XBox controller plugin. I though this is "the way to go"

File: Code/Samples/Asteroids/GameState.cpp (hunk 4 above)
Comment (Ingrater) [reply Code/Samples/Asteroids/GameState.cpp]: You are right. Will wrap.

File: Code/ThirdParty/glfw/finalize.cmake (hunk 5 above)
Comment (Ingrater) [reply Code/ThirdParty/glfw/finalize.cmake:16]: GLFW creates two cmake targets. One called "glfw" and the other one called "update_mappings" (which generates some lookup tables). I move both of them into the Code/ThirdParty filter within the visual studio solution.

**Pull request merged**